src/prof_recent.c

```diff
@@ -495,7 +546,47 @@ prof_recent_alloc_dump_node(emitter_t *emitter, prof_recent_t *n) {
 	emitter_json_object_end(emitter);
 }
 
-#define PROF_RECENT_PRINT_BUFSIZE 4096
+static void
+prof_recent_alloc_merge_locked(tsd_t *tsd, prof_recent_t *last) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_dump_mtx);
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	prof_recent_t *dump_dummy = prof_recent_alloc_dummy_reserve;
+	assert(last != NULL && last->next == dump_dummy);
+	last->next = prof_recent_alloc_dummy->next;
+	assert(dump_dummy->next != dump_dummy);
+	prof_recent_alloc_dummy->next = dump_dummy->next;
+	dump_dummy->next = dump_dummy;
+}
+
+static ssize_t
+prof_recent_alloc_dump_finish_locked(tsd_t *tsd, prof_recent_t *dump_last,
+    ssize_t dump_count, ssize_t dump_max) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_dump_mtx);
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	prof_recent_alloc_assert_count(tsd);
+	ssize_t new_max = dump_max;
+	if (dump_count > 0) {
+		prof_recent_alloc_merge_locked(tsd, dump_last);
+		prof_recent_alloc_count += dump_count;
+		assert(dump_max == prof_recent_alloc_max_get(tsd));
+		if (dump_max != -1 && prof_recent_alloc_count > dump_max) {
+			new_max = prof_recent_alloc_count;
+			assert(new_max > dump_max);
+			ssize_t old_max = prof_recent_alloc_max_update(tsd,
+			    new_max);
```

Comment:
Let’s put the max adjustment logic together. Only merge the lists in _finish() is fine. Having the max_update and max_write separated feels wrong.

Reply:
See my main comment.